Index: dnsmasq-2.66test3/src/radv.c
===================================================================
--- dnsmasq-2.66test3.orig/src/radv.c	2012-12-21 17:53:15.000000000 +0100
+++ dnsmasq-2.66test3/src/radv.c	2012-12-22 11:57:11.991647847 +0100
@@ -435,6 +435,7 @@
 		  opt->len = 4;
 		  opt->prefix_len = prefix;
 		  /* autonomous only if we're not doing dhcp, always set "on-link" */
+		  do_slaac = 1; // FIXME: Always enable autonomous flag
 		  opt->flags = do_slaac ? 0xC0 : 0x80;
 		  opt->valid_lifetime = htonl(valid);
 		  opt->preferred_lifetime = htonl(preferred);
